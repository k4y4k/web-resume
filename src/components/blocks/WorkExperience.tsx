import dayjs from "dayjs";
import { graphql, useStaticQuery } from "gatsby";
import * as React from "react";
import ItemContainer from "../section/ItemContainer";
import SectionContainer from "../section/SectionContainer";

interface ExperienceItem {
  company: string;
  endDate: string;
  highlights?: string[];
  position: string;
  startDate: string;
  summary: string;
  organization?: string;
  link: string;
}

interface two {
  history: ExperienceItem[];
}

export const PureWorkExperience = ({ history }: two): JSX.Element => {
  return (
    <div data-testid="experience">
      <SectionContainer title="Experience">
        {history.map((el) => {
          return (
            <ItemContainer
              key={el.organization + el.position}
              summary={el.summary}
              fromDate={el.startDate}
              toDate={el.endDate}
              title={el.position}
              subtitle={el.company ?? el.organization}
              link={el.link}
            />
          );
        })}
      </SectionContainer>
    </div>
  );
};

export const WorkExperience = (): JSX.Element => {
  const data = useStaticQuery(graphql`
    {
      file(name: { eq: "data" }, extension: { eq: "json" }) {
        id
        childDataJson {
          work {
            startDate
            endDate
            company
            position
            summary
            link
          }
        }
      }
    }
  `);

  const { work } = data.file.childDataJson;

  const aggregated = [...work].sort((a, b) =>
    dayjs(a.startDate).isAfter(b.startDate) ? -1 : 1,
  );

  return <PureWorkExperience history={aggregated} />;
};

export default WorkExperience;
