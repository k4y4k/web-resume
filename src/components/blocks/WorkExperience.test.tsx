import * as React from "react";
import { WorkExperience, PureWorkExperience } from "./WorkExperience";
import { render, screen } from "@testing-library/react";
import { useStaticQuery } from "gatsby";

describe("<Experience />", () => {
  beforeAll(() =>
    (useStaticQuery as jest.Mock).mockReturnValue({
      file: {
        childDataJson: {
          work: [
            {
              startDate: "2013-01-01",
              endDate: "2014-01-01",
              company: "Company",
              position: "President",
              summary: "Description...",
            },
          ],
          volunteer: [
            {
              startDate: "2012-01-01",
              endDate: "2013-01-01",
              organization: "Organization",
              position: "Volunteer",
              summary: "Description...",
            },
            {
              startDate: "2016-03-07",
              endDate: "2017-02-06",
              organization: "Organization",
              position: "Volunteer",
              summary: "Description...",
            },
          ],
        },
      },
    }),
  );

  test("bails out on no data", () => {
    render(<PureWorkExperience history={[]} />);

    // still want the title
    expect(screen.getByTestId("experience")).toBeInTheDocument();

    expect(
      screen.queryByTestId("sectionItemContainer"),
    ).not.toBeInTheDocument();
  });

  test("renders array of >=1 experience items", () => {
    const data = [
      {
        company: "Company",
        position: "President",
        website: "http://company.com",
        startDate: "2013-01-01",
        endDate: "2014-01-01",
        summary: "Description...",
        highlights: ["Started the company"],
        link: "",
      },
    ];

    render(<PureWorkExperience history={data} />);

    const experienceList = screen.getAllByTestId("experience");

    expect(experienceList.length).toBe(1);
  });

  test("renders OK", () => {
    render(<WorkExperience />);

    expect(screen.getByTestId("experience")).toMatchSnapshot();
  });
});
