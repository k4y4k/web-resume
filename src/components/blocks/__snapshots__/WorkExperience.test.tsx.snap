// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Experience /> renders OK 1`] = `
<div
  data-testid="experience"
>
  <div
    class="p-4 py-2"
    data-testid="sectionContainer"
  >
    <h2
      class="font-mono text-lg font-bold uppercase text-orchid-600"
      data-testid="sectionTitle"
    >
      Experience
    </h2>
    <div
      data-testid="sectionItemContainer"
    >
      <h1
        class="mb-1 font-bold text-lg leading-5"
        data-testid="itemTitle"
      >
        President
      </h1>
      <div
        class="italic mb-1"
        data-testid="byline"
      >
        <span
          data-testid="itemSubtitle"
        >
          Company
        </span>
         | 
        <span
          data-testid="sectionItemDates"
        >
          2013
           - 
          2014
        </span>
      </div>
      <div
        data-testid="sectionItemDetails"
      >
        <p>
          Description...
        </p>
      </div>
    </div>
  </div>
</div>
`;
